### PR TITLE
Fix #2864: Optimizer performance regression.

### LINKED ISSUE #2864: Severe performance regression of the parallel optimizer
Fast-optimizing the testSuite in the 0.6.x branch takes forever; I don't even wait until the end. Reverting back before #2814 fixes the issue.

Our CI jobs seem to take much longer than they used to, as well, so I don't think it is linked to my machine.

At any time, if I ask for a thread dump, all optimizing threads are snapshots with a stack trace similar to the following:

<details>

<summary>Full stack trace</summary>

```
"ForkJoinPool-1-worker-17" #100 daemon prio=5 os_prio=0 tid=0x00007f622400f000 nid=0x2c89 runnable [0x00007f622befb000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:48)
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:48)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:48)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:716)
	at scala.collection.parallel.immutable.LazyParVectorCombiner$$anonfun$result$1.apply(ParVector.scala:127)
	at scala.collection.parallel.immutable.LazyParVectorCombiner$$anonfun$result$1.apply(ParVector.scala:126)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.result(ParVector.scala:126)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.result(ParVector.scala:105)
	at scala.collection.parallel.Combiner$class.resultWithTaskSupport(Combiner.scala:89)
	at scala.collection.parallel.immutable.LazyParVectorCombiner.resultWithTaskSupport(ParVector.scala:105)
	at scala.collection.parallel.ParSeqLike$$anonfun$patch$3.apply(ParSeqLike.scala:218)
	at scala.collection.parallel.ParSeqLike$$anonfun$patch$3.apply(ParSeqLike.scala:218)
	at scala.collection.parallel.ParIterableLike$$anon$12$$anon$4.map(ParIterableLike.scala:306)
	at scala.collection.parallel.ParIterableLike$ResultMapping.leaf(ParIterableLike.scala:962)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$ResultMapping.tryLeaf(ParIterableLike.scala:956)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:165)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinTask.doJoin(ForkJoinTask.java:341)
	at scala.concurrent.forkjoin.ForkJoinTask.join(ForkJoinTask.java:673)
	at scala.collection.parallel.ForkJoinTasks$WrappedTask$class.sync(Tasks.scala:444)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:514)
	at scala.collection.parallel.ForkJoinTasks$class.executeAndWaitResult(Tasks.scala:492)
	at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:64)
	at scala.collection.parallel.ParSeqLike$class.patch(ParSeqLike.scala:217)
	at scala.collection.parallel.immutable.ParVector.patch(ParVector.scala:43)
	at scala.collection.parallel.ParSeqLike$class.$colon$plus(ParSeqLike.scala:250)
	at scala.collection.parallel.immutable.ParVector.$colon$plus(ParVector.scala:43)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.org$scalajs$core$tools$linker$frontend$optimizer$ConcurrencyUtils$AtomicAccOps$$append(ConcurrencyUtils.scala:47)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.$plus$eq$extension(ConcurrencyUtils.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$ParMethodImpl.registeredTo(ParIncOptimizer.scala:191)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.registerAskBody(GenIncOptimizer.scala:852)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl$Optimizer.getMethodBody(GenIncOptimizer.scala:949)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl$Optimizer.getMethodBody(GenIncOptimizer.scala:943)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inline(OptimizerCore.scala:1844)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformApplyStatic$1.apply(OptimizerCore.scala:1611)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformApplyStatic$1.apply(OptimizerCore.scala:1597)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1$$anonfun$apply$11.apply(OptimizerCore.scala:748)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1$$anonfun$apply$11.apply(OptimizerCore.scala:747)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs(OptimizerCore.scala:753)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1.apply(OptimizerCore.scala:747)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExprs$1.apply(OptimizerCore.scala:746)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:806)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:787)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:509)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList(OptimizerCore.scala:2308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList$2$$anonfun$apply$45.apply(OptimizerCore.scala:2294)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList$2$$anonfun$apply$45.apply(OptimizerCore.scala:2286)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBodyList(OptimizerCore.scala:2327)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1$$anonfun$apply$43.apply(OptimizerCore.scala:2217)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1$$anonfun$apply$43.apply(OptimizerCore.scala:2209)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings$1.apply(OptimizerCore.scala:3939)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings$1.apply(OptimizerCore.scala:3933)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDefs(OptimizerCore.scala:3966)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBindings(OptimizerCore.scala:3940)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1.apply(OptimizerCore.scala:2218)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$inlineClassConstructorBody$1.apply(OptimizerCore.scala:2189)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:567)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformExpr(OptimizerCore.scala:288)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$11.apply(OptimizerCore.scala:352)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$11.apply(OptimizerCore.scala:317)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon(OptimizerCore.scala:1117)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1050)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1048)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:806)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:787)
	at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
	at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:4149)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:355)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:728)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:735)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:402)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:276)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:728)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:735)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:308)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformIsolatedBody(OptimizerCore.scala:3783)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.liftedTree1$1(OptimizerCore.scala:135)
	at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:134)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.process(GenIncOptimizer.scala:935)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:88)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:87)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:975)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:972)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

</details>

Note in particular this portion:
```
	at scala.collection.parallel.immutable.ParVector.$colon$plus(ParVector.scala:43)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.org$scalajs$core$tools$linker$frontend$optimizer$ConcurrencyUtils$AtomicAccOps$$append(ConcurrencyUtils.scala:47)
	at org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils$AtomicAccOps$.$plus$eq$extension(ConcurrencyUtils.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$ParMethodImpl.registeredTo(ParIncOptimizer.scala:191)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.registerAskBody(GenIncOptimizer.scala:852)
```
It seems that `ParVector.:+` is an absolute performance killer, that should be avoided at all cost.

### PR DESCRIPTION
**Pull request by gzm0**
This essentially reverts commit ec3c3bd685ff32f9fbbe4e19db085cb651babf8c modulo:

- The build file changes to include the par-collections.
- An additional import to get `.toParArray` back.

Performance on testSuite/test:fastOptJS on
MacBook Air, 2.2 GHz Intel Core i7, 8 GB 1600 MHz DDR3

tl;dr: optimizer is 10x faster after the revert.

Before:

Linker: Check Infos: 5253899 us
Linker: Compute reachability: 2527607 us
Linker: Assemble LinkedClasses: 429050 us
Linker: Check IR: 2320856 us
Basic Linking: 10591499 us
Inc. optimizer: Batch mode: true
Inc. optimizer: Incremental part: 39471999 us
Inc. optimizer: Optimizing 35061 methods.
Inc. optimizer: Optimizer part: 348688137 us
Inc. optimizer: 388289830 us
Refiner: Compute reachability: 382626 us
Refiner: Assemble LinkedClasses: 82460 us
Refiner: 476345 us
Emitter: Class tree cache stats: reused: 2785 -- invalidated: 2975
Emitter: Method tree cache stats: resued: 0 -- invalidated: 24570
Emitter (write output): 6734101 us
Global IR cache stats: reused: 159 -- invalidated: 2187 -- trees read: 5747

After:

Linker: Check Infos: 5480910 us
Linker: Compute reachability: 3199914 us
Linker: Assemble LinkedClasses: 529704 us
Linker: Check IR: 2257044 us
Basic Linking: 11544069 us
Inc. optimizer: Batch mode: true
Inc. optimizer: Incremental part: 746521 us
Inc. optimizer: Optimizing 35061 methods.
Inc. optimizer: Optimizer part: 37023884 us
Inc. optimizer: 37889433 us
Refiner: Compute reachability: 380839 us
Refiner: Assemble LinkedClasses: 68282 us
Refiner: 463019 us
Emitter: Class tree cache stats: reused: 2785 -- invalidated: 2975
Emitter: Method tree cache stats: resued: 0 -- invalidated: 24570
Emitter (write output): 11078321 us
Global IR cache stats: reused: 0 -- invalidated: 2346 -- trees read: 5747

### CODE REVIEW

**Comment by gzm0** [comment]
I manually verified that this still passes:

```
> set resolvers in Global += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"
> ++2.13.0-pre-e2a2cba
> tools/compile
[success] Total time: 74 s, completed Apr 6, 2017 8:39:50 PM
```

**Comment by sjrd** [comment]
Jenkins, retest this please
